### PR TITLE
Added HIP support to VTK-m package

### PR DESCRIPTION
**Pull request by robertmaynard**

### CODE REVIEW

File: var/spack/repos/builtin/packages/vtk-m/package.py [hunk 1]
```diff
@@ -40,32 +40,54 @@ class VtkM(CMakePackage, CudaPackage):
...
-    variant("openmp", default=(sys.platform != 'darwin'), description="build openmp support")
     variant("rendering", default=True, description="build rendering support")
-    variant("tbb", default=(sys.platform == 'darwin'), description="build TBB support")
     variant("64bitids", default=False,
             description="enable 64 bits ids")
 
+    # Device variants
+    variant("cuda", default=False, description="build cuda support")
+    variant("openmp", default=(sys.platform != 'darwin'), description="build openmp support")
+    variant("tbb", default=(sys.platform == 'darwin'), description="build TBB support")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/vtk-m/package.py:55]: Do these two conflict or is it possible to build both/neither?

**Review by adamjstewart — COMMENTED** [review]
@kmorel @vicentebolea may also want to review

File: var/spack/repos/builtin/packages/vtk-m/package.py (hunk 1 above)
Comment (robertmaynard) [reply var/spack/repos/builtin/packages/vtk-m/package.py:55]: It is possible to build both/neither . Currently we support any number of CPU backends, but only 1 GPU backend.

**Review by adamjstewart — APPROVED** [review]

**Comment by adamjstewart** [comment]
Flake8:
```
var/spack/repos/builtin/packages/vtk-m/package.py:58: [E265] block comment should start with '# '
var/spack/repos/builtin/packages/vtk-m/package.py:68: [E303] too many blank lines (2)
var/spack/repos/builtin/packages/vtk-m/package.py:76: [E303] too many blank lines (2)
var/spack/repos/builtin/packages/vtk-m/package.py:168: [E501] line too long (84 > 79 characters)
```

File: var/spack/repos/builtin/packages/vtk-m/package.py [hunk 2]
```diff
@@ -40,32 +40,54 @@ class VtkM(CMakePackage, CudaPackage):
...
+    variant("hip", default=False, description="build hip support")
+
+    #it doesn't look like spack has a amd gpu abstraction
```
Comment (vicentebolea) [inline var/spack/repos/builtin/packages/vtk-m/package.py]:
```suggestion
    # it doesn't look like spack has a amd gpu abstraction
```

File: var/spack/repos/builtin/packages/vtk-m/package.py [hunk 3]
```diff
@@ -40,32 +40,54 @@ class VtkM(CMakePackage, CudaPackage):
...
+    amdgpu_targets = (
+        'gfx701', 'gfx801', 'gfx802', 'gfx803',
+        'gfx900', 'gfx906', 'gfx908', 'gfx1010',
+        'gfx1011', 'gfx1012'
+    )
+    variant('amdgpu_target', default='none', multi=True, values=amdgpu_targets)
+    conflicts("+hip", when="amdgpu_target=none")
+
+
```
Comment (vicentebolea) [inline var/spack/repos/builtin/packages/vtk-m/package.py]:
```suggestion
```

File: var/spack/repos/builtin/packages/vtk-m/package.py [hunk 4]
```diff
@@ -40,32 +40,54 @@ class VtkM(CMakePackage, CudaPackage):
...
     depends_on("cmake@3.12:", type="build")         # CMake >= 3.12
+    depends_on("cmake@3.18:", when="+hip")          # CMake >= 3.18
+
+    depends_on('cuda@10.2.0:', when='+cuda')
     depends_on("tbb", when="+tbb")
-    depends_on("cuda", when="+cuda")
     depends_on("mpi", when="+mpi")
 
+
```
Comment (vicentebolea) [inline var/spack/repos/builtin/packages/vtk-m/package.py]:
```suggestion
```

File: var/spack/repos/builtin/packages/vtk-m/package.py [hunk 5]
```diff
@@ -122,6 +124,51 @@ def cmake_args(self):
...
+                    options.append("-DVTKm_CUDA_Architecture=volta")
+            else:
+                options.append("-DVTKm_ENABLE_CUDA:BOOL=OFF")
+
+            # hip support
+            if "+hip" in spec:
+                options.append("-DVTKm_ENABLE_HIP:BOOL=ON")
+
+                archs = ",".join(self.spec.variants['amdgpu_target'].value)
+                options.append("-DCMAKE_HIP_ARCHITECTURES:STRING={0}".format(archs))
```
Comment (vicentebolea) [inline var/spack/repos/builtin/packages/vtk-m/package.py]:
```suggestion
                options.append("-DCMAKE_HIP_ARCHITECTURES:STRING={0}"
                    .format(archs))
```

**Review by vicentebolea — APPROVED** [review]

**Pull request merged**